### PR TITLE
Trajectory replay move

### PR DESCRIPTION
**Pull request by rc83**
# Description

* Refactor XTCReader and XTCWriter
* Add ReplayMove (stub)

A preview of mentioned features. Comments welcome.

The external xdrfile library has had to be patched. I have placed patches in `cmake/patches` and changed `CMakeLists.txt` accordingly.

## Checklist

- [ ] `make test` passes with no errors
- [ ] the source code is well documented
- [ ] new functionality includes unittests
- [ ] the user-manual has been updated to reflect the changes
- [ ] I have installed the provided commit hook to auto-format commits (requires `clang-format`):

  ``` bash
  ./scripts/git-pre-commit-format install
  ```

- [ ] code naming scheme follows the convention:

  Style        | Elements
  ------------ | ---------------------
  `PascalCase` | classes, namespaces
  `camelCase`  | functions
  `snake_case` | variables

### CODE REVIEW

File: src/io.cpp [hunk 1]
```diff
@@ -592,6 +523,231 @@ void FormatGRO::save(const std::string &filename, const Space &spc) {
...
+        frame = *xtc_frame;
+    }
+    return is_ok;
+}
+
+// ========== XTCWriter ==========
+
+XTCWriter::XTCWriter(const std::string &filename) : filename(filename) {
+    xdrfile = XDRfile::xdrfile_open(filename.c_str(), "w");
+    if (xdrfile == nullptr) {
```
Comment (mlund) [inline src/io.cpp:708]: `if (!xdrfile) { ... }`

File: src/io.h [hunk 2]
```diff
@@ -167,83 +167,363 @@ class FormatGRO {
...
+     * @param[in] begin  input iterator with coordinates in nanometers
+     * @param[in] end  input iterator's end
+     * @param[in] offset  offset in nanometers to add to all coordinates upon conversion
+     */
+    template <class begin_iterator, class end_iterator>
+    void importCoordinates(begin_iterator begin, end_iterator end, const Point &offset) const {
+        // setNumberOfAtoms(std::distance(begin, end));
+        // comparison of i is probably faster than prior call of std::distance
+        int i = 0;
+        for (auto coordinates_it = begin; coordinates_it != end; ++coordinates_it) {
```
Comment (mlund) [inline src/io.h:309]: use `std::transform` and `std::begin(xtc_coordinates.get())`. Throw can be moved outside loop.

File: src/io.h [hunk 3]
```diff
@@ -167,83 +167,363 @@ class FormatGRO {
...
+     * @brief Exports and converts atomic coordinates. Offset is subtracted to all coordinates to account different
+     * coordinates' origin.
+     * @tparam begin_iterator
+     * @tparam end_iterator
+     * @param[out] begin  output iterator with coordinates in nanometers
+     * @param[out] end  output iterator's end
+     * @param[in] offset  offset in nanometers to subtract from all coordinates upon conversion
+     */
+    template <class begin_iterator, class end_iterator>
+    void exportCoordinates(begin_iterator begin, end_iterator end, const Point &offset) const {
```
Comment (mlund) [inline src/io.h:349]: use `std::transform` and `std::begin(xtc_coordinates.get())`. Throw can be moved outside loop.

File: src/io.h [hunk 4]
```diff
@@ -167,83 +167,363 @@ class FormatGRO {
...
+ * C function calls.
+ *
+ * The frames can be provided as a TrajectoryFrame structure or as a list of positions in an input iterator. The class
+ * is responsible for I/O operations, not data conversion.
+ */
+class XTCWriter {
+    int return_code = XDRfile::exdrOK;   //!< last return code of a C function
+    XDRfile::XDRFILE *xdrfile = nullptr; //!< file handle
+    //! data structure for C functions; the number of coordinates is immutable
+    std::shared_ptr<XTCTrajectoryFrame> xtc_frame = nullptr;
```
Comment (mlund) [inline src/io.h:472]: `= nullptr` is redundant

File: src/move.cpp [hunk 5]
```diff
@@ -83,6 +83,30 @@ void to_json(json &j, const Movebase &m) {
...
+        } else {
+            // nothing to do, simulation shall stop
+            end_of_trajectory = true;
+            mcloop_logger->warn("No more frames to read from {}. Running on empty.", reader->filename);
+        }
+    }
+}
+
+double ReplayMove::bias(Change &, double, double) {
+    return -1e12; // always accept
```
Comment (mlund) [inline src/move.cpp:107]: rough

File: src/io.cpp [hunk 6]
```diff
@@ -592,6 +523,231 @@ void FormatGRO::save(const std::string &filename, const Space &spc) {
...
+
+XTCTrajectoryFrame::XTCTrajectoryFrame(const TrajectoryFrame &frame) {
+    initNumberOfAtoms(frame.coordinates.size());
+    importFrame(frame);
+}
+
+void XTCTrajectoryFrame::operator=(const TrajectoryFrame &frame) {
+    if (frame.coordinates.size() != number_of_atoms) {
+        // to avoid mistakes, the number_of_atoms is immutable
+        std::runtime_error("wrong number of particles to be assign into the XTC frame");
```
Comment (mlund) [inline src/io.cpp:538]: insert `throw`. Perhaps move inline comments into main doxygen code block

File: src/io.cpp [hunk 7]
```diff
@@ -592,6 +523,231 @@ void FormatGRO::save(const std::string &filename, const Space &spc) {
...
+    exportBox(frame.box);
+    exportCoordinates(frame.coordinates, 0.5 * frame.box);
+}
+
+void XTCTrajectoryFrame::exportTimestamp(int &step, float &time) const {
+    step = xtc_step;
+    time = xtc_time * 1.0_ps;
+}
+
+void XTCTrajectoryFrame::exportBox(Point &box) const {
```
Comment (mlund) [inline src/io.cpp:588]: add `/** @trows ... */` block w. explanation

File: src/io.h [hunk 8]
```diff
@@ -167,83 +167,363 @@ class FormatGRO {
...
+     * @tparam begin_iterator
+     * @tparam end_iterator
+     * @param[in] box  dimensions of the cubic box (xyz)
+     * @param[in] coordinates_begin  input iterator with coordinates (not particles)
+     * @param[in] coordinates_end  input iterator's end
+     */
+    template <class begin_iterator, class end_iterator>
+    void writeNext(const Point &box, begin_iterator coordinates_begin, end_iterator coordinates_end) {
+        if (xtc_frame == nullptr) {
+            int number_of_atoms = std::distance(coordinates_begin, coordinates_end);
```
Comment (mlund) [inline src/io.h:507]: int --> auto

File: src/move.h [hunk 9]
```diff
@@ -48,6 +49,27 @@ class Movebase {
...
 
+/**
+ * @brief Replay simulation from a trajectory
+ *
+ * Particles' positions are updated in every step based on coordinates read from the trajectory. Currently only
+ * XTCReader is supported.
+ */
+class ReplayMove : public Movebase {
+    Space &spc;                                  //!< space to operate on
+    std::shared_ptr<XTCReader> reader = nullptr; //!< trajectory reader
```
Comment (mlund) [inline src/move.h:60]: `= nullptr` redundant

File: src/space.h [hunk 10]
```diff
@@ -100,8 +100,8 @@ class Space {
     json info();
 
     //! Iterable range of all particle positions
-    auto positions() const {
-        return ranges::cpp20::views::transform(p, [](auto &i) -> const Point & { return i.pos; });
+    auto positions() {
```
Comment (mlund) [inline src/space.h:103]: keep both const and non-const version of function?

**Review by mlund — APPROVED** [review]
Nice! Please see minor comments.

File: src/move.cpp [hunk 11]
```diff
@@ -234,6 +258,8 @@ Propagator::Propagator(const json &j, Space &spc, Energy::Hamiltonian &pot, MPI:
                     _moves.emplace_back<Move::QuadrantJump>(spc);
                 else if (it.key() == "cluster")
                     _moves.emplace_back<Move::Cluster>(spc);
+                else if (it.key() == "replay")
```
[the remark this answers is not in the document]
Comment (rc83) [reply src/move.cpp:261]: done

File: src/io.cpp [hunk 12]
```diff
@@ -592,6 +523,231 @@ void FormatGRO::save(const std::string &filename, const Space &spc) {
...
+    XTCMatrix xtc_box_matrix = XTCMatrix::Zero();
+    // only XYZ dimensions in nanometers on diagonal, as floats
+    xtc_box_matrix.diagonal() = (box / 1.0_nm).cast<XTCFloat>();
+    // copy underlaying eigen structure (1D array, row-major) to the C-style 2D array
+    std::copy(xtc_box_matrix.data(), xtc_box_matrix.data() + DIM * DIM, &(xtc_box[0][0]));
+}
+
+void XTCTrajectoryFrame::importCoordinates(const PointVector &coordinates, const Point &offset) {
+    // setNumberOfAtoms(coordinates.size());
+    if (coordinates.size() != number_of_atoms) {
```
[the remark this answers is not in the document]
Comment (rc83) [reply src/io.cpp:565]: The size of the `xtc_coordinates` as raw C structure is determined by `number_of_atoms`. Memory is allocate and `number_of_atoms` set together in `initNumberOfAtoms()`. This is also the only place which sets the `number_of_atoms`.

File: src/io.cpp [hunk 13]
```diff
@@ -592,6 +523,231 @@ void FormatGRO::save(const std::string &filename, const Space &spc) {
...
+void XTCTrajectoryFrame::exportTimestamp(int &step, float &time) const {
+    step = xtc_step;
+    time = xtc_time * 1.0_ps;
+}
+
+void XTCTrajectoryFrame::exportBox(Point &box) const {
+    XTCMatrix xtc_box_matrix = Eigen::Map<const XTCTrajectoryFrame::XTCMatrix>(&(xtc_box[0][0]));
+    if (xtc_box_matrix.diagonal().asDiagonal().toDenseMatrix() != xtc_box_matrix) {
+        throw std::runtime_error("cannot load non-orthogonal box");
+    };
```
[the remark this answers is not in the document]
Comment (rc83) [reply src/io.cpp:592]: fixed

File: src/io.cpp (hunk 1 above)
Comment (rc83) [reply src/io.cpp:708]: fixed

File: src/io.h (hunk 4 above)
Comment (rc83) [reply src/io.h:472]: removed for the shared pointer

File: src/io.cpp (hunk 6 above)
Comment (rc83) [reply src/io.cpp:538]: both done

File: src/io.cpp (hunk 7 above)
Comment (rc83) [reply src/io.cpp:588]: done

File: src/io.h (hunk 8 above)
Comment (rc83) [reply src/io.h:507]: done

File: src/space.h (hunk 10 above)
Comment (rc83) [reply src/space.h:103]: done

File: src/move.h (hunk 9 above)
Comment (rc83) [reply src/move.h:60]: fixed

File: src/io.cpp [hunk 14]
```diff
@@ -592,6 +523,231 @@ void FormatGRO::save(const std::string &filename, const Space &spc) {
...
+    std::copy(xtc_box_matrix.data(), xtc_box_matrix.data() + DIM * DIM, &(xtc_box[0][0]));
+}
+
+void XTCTrajectoryFrame::importCoordinates(const PointVector &coordinates, const Point &offset) {
+    // setNumberOfAtoms(coordinates.size());
+    if (coordinates.size() != number_of_atoms) {
+        // to avoid mistakes, the number_of_atoms is immutable
+        throw std::runtime_error("wrong number of particles to be saved in the XTC frame");
+    }
+    for (auto i = 0; i < number_of_atoms; ++i) {
```
[the remark this answers is not in the document]
Comment (rc83) [reply src/io.cpp:569]: It seems `std::begin` cannot be used like that for 2D C-arrays. I think that the for loop is comprehensible when accessing 2D C-arrays. Won’t change.

File: src/io.cpp [hunk 15]
```diff
@@ -592,6 +523,231 @@ void FormatGRO::save(const std::string &filename, const Space &spc) {
...
+void XTCTrajectoryFrame::exportBox(Point &box) const {
+    XTCMatrix xtc_box_matrix = Eigen::Map<const XTCTrajectoryFrame::XTCMatrix>(&(xtc_box[0][0]));
+    if (xtc_box_matrix.diagonal().asDiagonal().toDenseMatrix() != xtc_box_matrix) {
+        throw std::runtime_error("cannot load non-orthogonal box");
+    };
+    box = Point(xtc_box_matrix.diagonal().cast<double>() * 1.0_nm);
+}
+
+void XTCTrajectoryFrame::exportCoordinates(PointVector &coordinates, const Point &offset) const {
+    // coordinates.resize(number_of_atoms);
```
[the remark this answers is not in the document]
Comment (rc83) [reply src/io.cpp:597]: done

File: src/io.h (hunk 2 above)
Comment (rc83) [reply src/io.h:309]: Does not look as straightforward with 2D C-arrays. Keeping the current code.

File: src/io.h (hunk 3 above)
Comment (rc83) [reply src/io.h:349]: Does not look as straightforward with 2D C-arrays. Keeping the current code.

File: cmake/ExternalTools.cmake [hunk 16]
```diff
@@ -222,7 +222,8 @@ ExternalProject_Add(
     PREFIX "${CMAKE_CURRENT_BINARY_DIR}/_deps"
     URL "https://github.com/wesbarnett/libxdrfile/archive/2.1.2.tar.gz"
     URL_MD5 ee114404b4a01613b2f0167a2ad92536
-    PATCH_COMMAND echo "add_library(xdrfile-static STATIC \${SRCS})" >> CMakeLists.txt
```
[the remark this answers is not in the document]
Comment (rc83) [reply cmake/ExternalTools.cmake:225]: done

**Pull request merged**